packages/theia-integration/src/node/java-socket-glsp-server.ts

```diff
@@ -0,0 +1,124 @@
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import * as fs from "fs";
+import { injectable, postConstruct } from "inversify";
+import * as net from "net";
+import { createSocketConnection, IConnection } from "vscode-ws-jsonrpc/lib/server";
+
+import { BaseGLSPServerContribution, GLSPServerLaunchOptions } from "./glsp-server-contribution";
+
+export const START_UP_COMPLETE_MSG = "[GLSP-Server]:Startup completed";
+export interface JavaSocketServerLaunchOptions extends GLSPServerLaunchOptions {
+    /** Path to the location of the jar file that should be launched as process */
+    jarPath: string,
+    /** Port on which the server should listen for new client connections */
+    serverPort: number,
+    /** Additional arguments that should be passed when starting the server process. */
+    additionalArgs?: string[]
+}
+
+export namespace JavaSocketServerLaunchOptions {
+    /** Default values for {@link JavaGLSPServerLaunchOptions }**/
+    export function createDefaultOptions(): JavaSocketServerLaunchOptions {
+        return <JavaSocketServerLaunchOptions>{
+            ...GLSPServerLaunchOptions.createDefaultOptions(),
+            jarPath: "",
+            serverPort: NaN
+        };
+    }
+
+    /**
+    * Utility function to partially set the launch options. Default values (from 'defaultOptions') are used for
+    * options that are not specified.
+    * @param options (partial) launch options that should be extended with default values (if necessary)
+    */
+    export function configure(options?: Partial<JavaSocketServerLaunchOptions>): JavaSocketServerLaunchOptions {
+        return options ?
+            <JavaSocketServerLaunchOptions>{
+                ...createDefaultOptions(),
+                ...options
+            } : createDefaultOptions();
+    }
+}
+
+/**
+ *  A reusable base implementation for {@link GLSPServerContribution}s that are using a socket connection to communicate
+ *  with a java-based GLSP server.
+ **/
+@injectable()
+export abstract class JavaSocketServerContribution extends BaseGLSPServerContribution {
+
+    protected resolveReady: (value?: void | PromiseLike<void> | undefined) => void;
+    onReady: Promise<void> = new Promise(resolve => this.resolveReady = resolve);
+    launchOptions: JavaSocketServerLaunchOptions;
+
+    @postConstruct()
+    protected initialize() {
+        if (this.createLaunchOptions) {
+            this.launchOptions = JavaSocketServerLaunchOptions.configure(this.createLaunchOptions());
+        }
+    }
+
+    abstract createLaunchOptions(): Partial<JavaSocketServerLaunchOptions>;
+
+    connect(clientConnection: IConnection): void {
+        this.connectToSocketServer(clientConnection);
+    }
+    async launch(): Promise<void> {
+        if (!this.launchOptions.multiClient || !this.isServerRunning()) {
```

Comment:
It seems like this method is only calls once the contribution starts (at least for the workflow case where we bind the contribution as a singleton), so I am not sure if the evaluation of the multiClient parameter ever is relevant because initially is the server is never running. It could be I am missing something so maybe you can explain that a bit ;-)

Reply:
You are absolutely right. The evaluation of the multiClient parameter is redundant here. I will remove it.